### PR TITLE
chore(all): run TS compiler as part of test-many

### LINKED ISSUE #12823: Run TypeScript compiler as part of pre-commit hook
We use esbuild to compile TS for a host of things, including all the tests.  esbuild is much faster, which really helps with our CI spend.  However, esbuild does not type check.  That means a PR would pass CI, then fail in `main` or a tag when we actually build the packages.  This has happened numerous time.  We should run `tsc --build` for the packages that need it somewhere before a PR’s CI run.  The pre-commit hook seems reasonable.  (For extra credit maybe check what files have been changed in the commit and only run tsc for those packages.)

┆Issue is synchronized with this [Jira Task](https://mozilla-hub.atlassian.net/browse/FXA-5063)

### PR DESCRIPTION
**Pull request by biancadanforth**
Because:
* We should be informed about TS errors before a patch merges into main.  

This commit:  
* Adds a new script to runs ~~on a pre-commit git hook~~ in the `test-many` job that ~~rebuilds~~ compiles any modified and dependent back-end TypeScript packages.
    
Closes #12823
    
Co-authored-by: Barry Chen <bchen@mozilla.com>

## Checklist

_Put an `x` in the boxes that apply_

- [x] My commit is GPG signed.
- [x] If applicable, I have modified or added tests which pass locally.
- [x] I have added necessary documentation (if appropriate).
- [x] I have verified that my changes render correctly in RTL (if appropriate).

## Other information (Optional)

This takes a long time if you modify e.g. `fxa-shared`, since 18 packages depend on it (granted, a few don’t have build commands), even with `yarn workspaces foreach` running some builds in parallel. I’m reluctant to land this as-is for this reason. Is there a better place where this can happen?

Time it took my machine to build `fxa-shared` and all the packages that depend on it:
```
real    1m34.764s
user    2m18.828s
sys    0m13.881s
```

Some options:
* Add a prompt, so devs can skip it for draft patches.
* Run it when a PR opens in CI, though this would increase our CircleCI usage by several minutes per PR.
* Audit workspace dependencies to make sure we're not building packages unnecessarily (this won't help much for widely shared packages like `fxa-shared`).

@bbangert , @dschom : What do you think?

### CODE REVIEW

File: _scripts/git-precommit-hook.sh [hunk 1]
```diff
@@ -0,0 +1,32 @@
+#!/bin/bash
+
+# 1. Get the list of all modified and dependent packages.
+# Cleans up yarn's output by removing everything before the package name, the first and the last line.
+PACKAGES_MODIFIED=`yarn workspaces list --since=HEAD^ -R | cut -d/ -f2 | sed '1d' | sed '$d'`
```
Comment (biancadanforth) [inline _scripts/git-precommit-hook.sh:5]: ## Test cases I tried

### Passed
* Only fxa root files modified
  * No building done
* Fxa root files and one package modified
  * Builds the modified package and any packages that depend on it successfully.
* One package modified
  * Builds the modified package and any packages that depend on it successfully.
* Multiple packages modified with overlapping packages that depend on them.
  * Builds the modified package and any packages that depend on it once each successfully.
* No files modified (e.g. `git commit --amend` to reword a commit)
  * No building done
* Multiple packages modified where one has a build error.
  * Error message logged to console and build logs accessible via `./artifacts/build-modified-and-dependent.log`.

### Failed
* Modified package with build error, but it is not staged.
  * No building done.
  * AFAICT [`yarn workspaces list`](https://yarnpkg.com/cli/workspaces/list) doesn't have a `--staged` option. I'm passing `HEAD^` as the git ref in the `--since` option, but this will pick up anything that has changed.
  * This is a less common scenario in my experience, but I'd like to get that working too if possible -- any ideas?

File: package.json [hunk 2]
```diff
@@ -6,7 +6,6 @@
   "scripts": {
     "preinstall": "_scripts/check-package-manager.sh",
     "authors": "git shortlog -s | cut -c8- | sort -f > AUTHORS",
-    "audit": "echo 'audit is currently unavailable'",
```
Comment (biancadanforth) [inline package.json:9]: The first commit is removing an unrelated noop run script.

It looks like the script was a way to reduce `npm audit`’s frequency (https://github.com/mozilla/fxa/pull/4831), but the root package.json file’s `audit` command is a noop. Is there something that should be happening here that isn’t? If it’s not a clear cut answer, I will remove this first commit entirely.

**Review by biancadanforth — COMMENTED** [review]
Leaving a couple of other questions for you inline.

Note: You can check `./artifacts/build-modified-and-dependent.log` for all build logs. I mention this file in the output if a build fails.

**Comment by chenba** [comment]
> Time it took my machine to build `fxa-shared` and all the packages that depend on it:
> 
> ```
> real    1m34.764s
> user    2m18.828s
> sys    0m13.881s
> ```

☹️ That's definitely long enough to be annoying.

> 
>     * Run it when a PR opens in CI, though this would increase our CircleCI usage by several minutes per PR.
> 

Since CI can't benefit from incremental builds and it runs in a VM much slower than your MBP, I'd expect it take quite a bit longer.  But I think it's the best option.

**Comment by biancadanforth** [comment]
> Since CI can't benefit from incremental builds and it runs in a VM much slower than your MBP, I'd expect it take quite a bit longer. But I think it's the best option.

Thanks Barry for your thoughts; I [asked Wil and Anna's opinion on doing this](https://mozilla-hub.atlassian.net/browse/FXA-5063?focusedCommentId=554322) in the Jira ticket, as I expect in addition to being slower on the VMs (great point), it will cost us more.

**Comment by dschom** [comment]
Yes, I agree with @chenba. This amount of wait time seems like it could be annoying. Especially if you are working on something in fxa-shared that isn’t actually used by all 18 packages… Having it be part of the CI seems better to me.
 
Just thinking out loud, but the fact that making a small change to fxa-shared results in such a large-scale rebuild may be an indicator that fxa-shared is too large of package? 

Personally, I’d be glad to wait 2 minutes after making a change to fxa-shared/sentry if I knew it was protecting me from a botched release down the road. However, I’d be annoyed if I was just adding a field to a fxa-shared/db/dto that was only used by 2 out of the 18 packages, and yet I had to wait for everything to rebuild. Maybe it's time to break fxa-shared down into a series of smaller packages? I think there might be other benefits to this as well, such as clear dependencies, better partitioning of client/server libraries, and adherence to npm/unix 'small is beautiful' design philosophy.

**Comment by annasob** [comment]
I personally rather this be a pre-commit hook as opposed to a CI. Is there a way to omit the building of dependent packages. If you haven't changed anything in the code we should skip building it.

**Comment by chenba** [comment]
> I personally rather this be a pre-commit hook as opposed to a CI. Is there a way to omit the building of dependent packages. If you haven't changed anything in the code we should skip building it.

@annasob thanks for bringing up this point.  @biancadanforth mentioned that there was at least once where a auth-server build failed because of a change in fxa-shared (@biancadanforth please correct me if I've mis-represented something here).  I think that's why this patch is building the dependent packages.  However, thinking over it a bit, prompted by Anna's comment, I have a follow-up question... @biancadanforth even though the auth-server build failed, was the error from building fxa-shared?  If so, then simply building fxa-shared pre-commit would alert us to the error.

**Comment by biancadanforth** [comment]
@dschom @bbangert This is ready for you. I got this back to being fast enough after talking with Ben.

TL;DR: We don't need to `--build`, since we don't need the built files. We can simply `tsc --noEmit`. We also don't need to worry about front-end packages, since those are built as part of CI already.

Time for modifying `fxa-shared`:
```
real	0m2.346s
user	0m1.992s
sys	0m0.356s
```

File: _scripts/git-precommit-hook.sh [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+for package_modified in $PACKAGES_MODIFIED; do
+  if [[ " ${BACKEND_PACKAGES[@]} " =~ " ${package_modified} " ]]; then
+      INCLUDE_ARGS="$INCLUDE_ARGS --include ${package_modified}"
+  fi
+done
+IFS=$ORIGINAL_IFS
+INCLUDE_ARGS=`echo "$INCLUDE_ARGS" | xargs` # trim whitespace at beginning of string
+echo "building all modified and dependent backend TS packages..."
+# We don't need to write files to disk, so use --noEmit here instead of --build for speed.
+if ! `yarn workspaces foreach --verbose ${INCLUDE_ARGS} run tsc --noEmit > artifacts/build-affected-backend-packages.log`;
```
Comment (pdehaan) [inline _scripts/git-precommit-hook.sh:30]: Q: Should we add `--parallel`?

Comment (pdehaan) [reply _scripts/git-precommit-hook.sh:30]: Also, not sure if I'm missing something, but I don't think this is working for me locally.

I'm trying to hardcode the command in my Terminal; but it's very fast and suspiciously unverbose:
```sh
yarn --version # 3.1.1
yarn workspaces foreach --verbose --include fxa-event-broker --include fxa-admin-server --include fxa-graphql-api --include fxa-auth-server -p run tsc --noEmit
➤ YN0000: Done in 0s
```

But, if I change <kbd>run tsc --noEmit</kbd> to <kbd>run build</kbd> then it seems to work and run the scripts. So, looks like I can run scripts from package.json's, but not arbitrary commands:
```sh
yarn workspaces foreach --verbose --include fxa-event-broker --include fxa-admin-server --include fxa-graphql-api --include fxa-auth-server -p run build
➤ YN0000: [fxa-graphql-api]: Process exited (exit code 0), completed in 6s 339ms
➤ YN0000: [fxa-event-broker]: Process exited (exit code 0), completed in 7s 188ms
➤ YN0000: [fxa-admin-server]: Found 4 error(s).
➤ YN0000: [fxa-admin-server]: (.....a few errors.....)
➤ YN0000: [fxa-admin-server]: Process exited (exit code 1), completed in 14s 781ms
➤ YN0000: [fxa-auth-server]: (node:61006) [DEP0147] DeprecationWarning: In future versions of Node.js, fs.rmdir(path, { recursive: true }) will be removed. Use fs.rm(path, { recursive: true }) instead
➤ YN0000: [fxa-auth-server]: (Use `node --trace-deprecation ...` to show where the warning was created)
➤ YN0000: [fxa-auth-server]: Process exited (exit code 0), completed in 20s 376ms
➤ YN0000: Done in 20s 386ms
```

Comment (biancadanforth) [reply _scripts/git-precommit-hook.sh:30]: Ah thanks; you're right. My husband was rushing me for dinner. I'm not trying to blame him, but it's definitely his fault. I'll update the script.

Comment (biancadanforth) [reply _scripts/git-precommit-hook.sh]: > Q: Should we add `--parallel`?

I don't think so, because some packages depend on other packages, so we want to compile those independent packages first (e.g. `fxa-shared` should be compiled before `fxa-auth-server`, since the latter depends on it).

Although... I'm not exactly sure how `--noEmit` works though, since for `--build`, we'd get a physical built file that can be included in the next package's compilation. 🤔  I'll see if I can find out.

**Comment by biancadanforth** [comment]
Okay modifying `fxa-shared` is now at 37s. Sorry for the confusion with the 2s.
```
real	0m36.970s
user	0m49.033s
sys	0m3.737s
```
This still seems too slow to me for a precommit hook. I was hoping for 10s or less. 😖

**Comment by biancadanforth** [comment]
> @biancadanforth mentioned that there was at least once where a auth-server build failed because of a change in fxa-shared (@biancadanforth please correct me if I've mis-represented something here).

The issue I had in mind was that I removed a type in `fxa-shared` that was used in other packages like `fxa-support-panel`. `fxa-shared` built correctly, but `fxa-support-panel` failed to build, because I forgot to remove the import/usage of the deleted type in `fxa-support-panel`. See this [CI failure](https://app.circleci.com/pipelines/github/mozilla/fxa/30437/workflows/dd5e5fac-2b28-4dbe-8a7e-5ee094e27aa1/jobs/335250/steps) and this [PR](https://github.com/mozilla/fxa/pull/12974) to fix it.

So the issue is we can make a change to one package, but if we don't compile that package and any package that depends on it, we can miss some errors that we only currently discover after the PR is merged into `main`.

> @biancadanforth even though the auth-server build failed, was the error from building fxa-shared? If so, then simply building fxa-shared pre-commit would alert us to the error.

No, see above.

Thank you both for your thoughtful input here. This is turning out to be a little tricksier than I thought. :-)

File: _scripts/git-precommit-hook.sh (hunk 3 above)
Comment (biancadanforth) [reply _scripts/git-precommit-hook.sh]: I just ran it with `--parallel`, and it looks like it still ran and completed `fxa-shared` first before parallelizing the other packages. So seems like it'd work? I got 24s down from 37s.

Unfortunately, despite [yarn's docs](https://yarnpkg.com/cli/workspaces/foreach) stating the opposite, the output seems to be interlaced[1]:
> If -p,--parallel and -i,--interlaced are both set, Yarn will print the lines from the output as it receives them. If -i,--interlaced wasn't set, it would instead buffer the output from each process and print the resulting buffers only after their source processes have exited.

[1]:
```
➤ YN0000: [fxa-shared]: Process started
➤ YN0000: [fxa-shared]: Process exited (exit code 0), completed in 9s 418ms
➤ YN0000: [fxa-admin-server]: Process started
➤ YN0000: [fxa-auth-server]: Process started
➤ YN0000: [fxa-event-broker]: Process started
➤ YN0000: [fxa-support-panel]: Process started
➤ YN0000: [fxa-support-panel]: Process exited (exit code 0), completed in 3s 18ms
➤ YN0000: [fxa-event-broker]: Process exited (exit code 0), completed in 6s 397ms
➤ YN0000: [fxa-admin-server]: Process exited (exit code 0), completed in 8s 802ms
➤ YN0000: [fxa-auth-server]: Process exited (exit code 0), completed in 10s 734ms
➤ YN0000: [fxa-graphql-api]: Process started
➤ YN0000: [fxa-graphql-api]: Process exited (exit code 0), completed in 3s 886ms
➤ YN0000: Done in 24s 52ms
```

**Comment by biancadanforth** [comment]
I still think~25s is too slow for a pre-commit hook, so I'm moving this back to a draft.

**Comment by dschom** [comment]
@biancadanforth I just exercised this a bit and it works nicely aside for the wait time. I see you've put it back into draft status, but the one piece of feedback is that I’d like to see the workspaces that were just built printed out if possible. Doing that _might_ make me feel better about the wait time. 
 
Looking through the comments above, there are also a couple other thoughts I might add. 

- It is not just changes in fxa-shared that cause the problem. I [dug up this fix](https://github.com/mozilla/fxa/commit/cae1f613d8fe37e187a2be0e70781c3eb9bd18b2) for a release that failed when I was release owner. You can see here fxa-shared is not the issue. The issue is that the typescript compiler and esbuild had different behaviors concerning require and import. There have been others similar failures this is just one I recall off hand.

- It is interesting that some of the code is compiled during testing. Should we maybe just do this for all the top level packages during CI testing?  That way the build gets spread across each package and maybe it isn’t as noticeable in the ci? 

- I am not sure I fully understand the aversion to doing a typescript build in in the CI. Is it cpu credits, or time, or both? If it’s time, we can run it in parallel, just as other jobs do. If it’s cpu credits, did we measure the savings when we switched to esbuild? Even though esbuild does seem really efficient / fast, much of this comes from smart parallelization. It’s possible that parallelization results in faster code, but doesn’t actually save you any compute costs, i.e. the number of operations is the same, but just spread across many cores. This is total conjecture and it's possible that esbuild is just way more effiencient in everyway. I guess really I am just curious about any measurements we might have.

- I am curious if there is a reason we don't use esbuild in prod? Maybe this is another way to alleviate some of the problems this PR is trying to address. It seems like even though it's a v0, it is [production ready](https://esbuild.github.io/faq/#production-readiness)? Whenever testing state and production state differ, which is always bound to happen, errors can creep in. Using esbuild for testing / dev, and then using a typescript build for prod to me is a difference, although it seems like it hasn't caused too many problems so far.

**Comment by chenba** [comment]
> I still think~25s is too slow for a pre-commit hook

Agreed.

> 
>     * It is interesting that some of the code is compiled during testing. 
> 

I feel I'm at least partly at fault for making some rather general statements in the issue I filed, leading to some confusion/misunderstanding.  esbuild is the replacement for ts-node, which we started using when TS was introduced into the repo for some packages.  But for the React front ends, we've always relied on whatever toolchain Create React App provided, so that was never changed because it wasn't using ts-node anyway.

>     * I am not sure I fully understand the aversion to doing a typescript build in in the CI. ...
> 

esbuild is faster because it's a compiled Go binary.  The TypeScript compiler (and Babel) is in JS/TS.  It's the compile step that's faster; I'm doubtful that it produces faster JavaScript.  (But it would be interesting to find out!)  Anyways I don't think we have an aversion of doing a TS build in CI; this could be PR where we do that. 😀 

>     * I am curious if there is a reason we don't use esbuild in prod? ...

I don't think there is a specific reason.  But the esbuild doc has a list: https://esbuild.github.io/content-types/#typescript-caveats (which I have not looked into).  We definitely don't want to skip the type checking step, so we still need to run a `tsc` build somewhere at least once (even if we don't emit JS).  So _if_ we want to use esbuild to build prod, this PR helps us to get there.  But I also don't have a great reason for switching to esbuild either.

> Whenever testing state and production state differ, which is always bound to happen, errors can creep in. Using esbuild for testing / dev, and then using a typescript build for prod to me is a difference, although it seems like it hasn't caused too many problems so far.

I certainly agree in principle.  (And you can also make the same argument for frontend code that's only uglified/bundled/optimized for prod.)  However I'm not too worry about issues so subtle that they sneak past the smoking tests and QA to prod.  But hey famous last words...

**Comment by biancadanforth** [comment]
> Just thinking out loud, but the fact that making a small change to fxa-shared results in such a large-scale rebuild may be an indicator that fxa-shared is too large of package?

That’s an interesting thought, though it's beyond the scope of this ticket. Do you want to file a spike and link it to this ticket to explore that more? It could help reduce unnecessary checks/tests among other benefits.

> It is interesting that some of the code is compiled during testing. Should we maybe just do this for all the top level packages during CI testing? That way the build gets spread across each package and maybe it isn’t as noticeable in the ci?

One option is to ensure we run `yarn build` (or a more efficient command like `tsc --noEmit` where possible) on any remaining TS packages that aren't already getting built in say the `test_many` job, which runs as part of the `test_pull_request` workflow.

Another option (Barry's original suggestion) is that I create a new job that runs this script in parallel with our existing jobs. `test-many` does not process each package in parallel like my current script does, so my sense is that both overall runtime and cpu credits are lower with the separate, parallel job. Not to mention it'd be easier to adapt the current script in this PR that way.

Unless there are any objections (cc @annasob , @clouserw ), I am going to try out this latter approach to see what the impact would be on our CI usage now that I've improved the performance by 73% (from 1m35s to 25s).

> I am curious if there is a reason we don't use esbuild in prod?
As Barry noted, `esbuild` doesn't type check. Their own [docs](https://esbuild.github.io/content-types/#typescript) state `tsc --noEmit` should be run in parallel. But that's probably only part of the answer. This decision predates me.

File: package.json [hunk 4]
```diff
@@ -46,8 +45,7 @@
   },
   "husky": {
     "hooks": {
-      "pre-commit": "lint-staged",
-      "pre-push": "_scripts/audit.sh",
+      "pre-commit": "lint-staged && _scripts/git-precommit-hook.sh",
```
Comment (pdehaan) [inline package.json:48]: Random question while I was trolling the Husky docs and .git/hooks/ directory... Do we want to use `pre-commit` or `pre-push`?
Not sure how [other] people use Git, but I know we've had variations of this argument over the past X years since some people commit a bunch of changes before pushing, and others it's more of a moot 1:1.

File: _scripts/git-precommit-hook.sh [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+    exit 0
+fi
+
+echo "modified packages found"
+
+# 3. Build all back-end TS packages in the list
+ORIGINAL_IFS=$IFS
+IFS=$'\n'
+# We don't need to worry about front-end packages that use TS, since they must build successfully for CI to pass.
+BACKEND_PACKAGES=( "fxa-auth-server" "fxa-event-broker" "fxa-graphql-api" "fxa-shared" "fxa-support-panel" "fxa-admin-server" )
```
Comment (pdehaan) [inline _scripts/git-precommit-hook.sh:19]: nit: Not sure if this list is in a special order, or if we should A-z it so "fxa-admin-server" is first so it's alphabetical.
Or, if they _are_ in a special build order, maybe a bonus comment/warning not to reorder them, lest we anger the build gods.

:+1: Also verified that each of these packages has a `compile` script (which we knew from the diffs below, but was mostly curious if other packages had the same command name as well):

```sh
git grep -n '"compile": ' packages/*/package.json | cat

packages/fxa-admin-server/package.json:8:    "compile": "tsc --noEmit",
packages/fxa-auth-server/package.json:13:    "compile": "tsc --noEmit",
packages/fxa-event-broker/package.json:9:    "compile": "tsc --noEmit",
packages/fxa-graphql-api/package.json:8:     "compile": "tsc --noEmit",
packages/fxa-shared/package.json:13:         "compile": "tsc --noEmit",
packages/fxa-support-panel/package.json:11:  "compile": "tsc --noEmit",
```

File: package.json (hunk 4 above)
Comment (biancadanforth) [reply package.json:48]: Thanks Peter; yeah I talked about it with a couple of other folks (Dan and Barry). Our sense was that our workflows tend to be 1:1 between commits and pushes, so it wouldn't make a big difference.

That said, my current plan is to move this out of a git hook and make a separate, parallel PR CI task, so it likely won't matter for this PR after all.

File: _scripts/git-precommit-hook.sh (hunk 5 above)
Comment (biancadanforth) [reply _scripts/git-precommit-hook.sh:19]: The order of the array doesn't matter, since `yarn workspaces foreach ... run compile` will order the processing based on the dependency graph. But I can make it alphabetical just to be a little easier to read/modify in the future. Thanks for the suggestion.

**Comment by dschom** [comment]
@chenba @biancadanforth Thanks for answering/entertaining my comments. I think I see the bigger picture a bit more clearly now.

BTW, I like the idea of running the `tsc --noEmit` in the CI. It's possible doing so could actually save circleci credits if we require that this passes before running other tests. 

Our success rate for test_pull_request in circleci is currently hovering around 40%. Therefore, anything we can do to prevent expensive (and often flaky) functional tests from running when we know there’s no point could save credits in the long term. @clouserw, @annasob thoughts?

**Review by biancadanforth — COMMENTED** [review]
@dschom , @bbangert : Third time is a charm? 😅 

It's basically the same script, but now with better logging and running as part of the `test-many` job in our `test_pull_request` workflow. In the worst case scenario modifying `fxa-shared` ([example from a test PR](https://app.circleci.com/pipelines/github/mozilla/fxa/31004/workflows/7f5c51e0-3418-4a85-a0cb-020a05cd6d07/jobs/340597/steps) (see the logs in the Artifacts tab)), it is adding a little over 1 minute to the PR build time.

I paired with Barry earlier, and we decided the best place for this check to live is in the `test-many` job, because:
* Creating a new job has some overhead, which can use a lot of CPU credits, so we hypothesize using an existing job is not only easier, but cheaper.
* A new job will only run in parallel if there are [enough containers in our CircleCI plan](https://circleci.com/docs2/2.0/faq#how-many-jobs-can-i-run-at-one-time).
  * I don't know where we stand here, but it's a complicating factor.
* `test-many` also runs on tags, which is another good place to perform this check, since for dot releases, we cherry-pick commits from `main` onto the train branch, which in almost all cases is an earlier version of `main`.

I added some comments in line.

I will roll these commits into 1 commit before merging. Keeping them here for now to preserve the review history (kind of, since i rebased a couple times).

File: _scripts/compile-backend-ts-services.sh [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+if [[ $AFFECTED_PACKAGES == "" ]]; then
+    echo "no affected services found"
+    exit 0
+fi
+echo "affected services found:"
+echo -e "$AFFECTED_PACKAGES" | sed '1d'
+
+# 4. Compile all back-end TS services in the list
+echo "compiling all modified and dependent backend TS services..."
+echo -e "\nNote: if fxa-shared was modified and has any TS errors, the script will exit before compiling other services, since all other backend TS services depend on fxa-shared."
```
[the remark this answers is not in the document]
Comment (chenba) [reply _scripts/compile-backend-ts-services.sh:42]: Couldn't hurt.  Thanks for following up.

**Comment by biancadanforth** [comment]
> BTW, I like the idea of running the `tsc --noEmit` in the CI. It's possible doing so could actually save circleci credits if we require that this passes before running other tests.

I can confirm that if this run task fails, the `test-many` job bails before it runs the subsequent tasks, but the other jobs in the workflow will continue unless/until they fail ([example run](https://app.circleci.com/pipelines/github/mozilla/fxa/31059/workflows/9956b726-cb7b-4f21-9333-88c5ebc7833e/jobs/341036) with a TS error in fxa-shared), since they don't `require` `test-many` in the config.

This could be changed in terms of stopping other jobs, but I would encourage us to consider that for a follow-up, is this current patch is still an improvement. I'm happy to file something if folks think it's a good idea to run `test-many` first, but that would also increase overall CI running time, since it would no longer run in parallel with the other jobs.
